### ISSUE TITLE
Errors trying to use functionality from groovy jar

### ISSUE THREAD

**Issue by collinpeters** [oracle/graal#4061]
**Describe the issue**
This is a follow-up to #3447. Specifically having issues trying to use functionality from within the groovy jar.

Reproduction repo: https://github.com/collinpeters/graalvm-groovy

I get this error when trying to use the groovy `AstBuilder` to load a `build.gradle` file:

```
Exception in thread "main" com.oracle.svm.core.jdk.UnsupportedFeatureError: No classes have been predefined during the image build to load from bytecodes at runtime.
        at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:87)
        at com.oracle.svm.core.hub.PredefinedClassesSupport.throwNoBytecodeClasses(PredefinedClassesSupport.java:76)
        at com.oracle.svm.core.hub.PredefinedClassesSupport.loadClass(PredefinedClassesSupport.java:130)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:269)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts.define(ClassLoaderForClassArtifacts.java:43)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts$1.run(ClassLoaderForClassArtifacts.java:76)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts$1.run(ClassLoaderForClassArtifacts.java:74)
        at java.security.AccessController.doPrivileged(AccessController.java:87)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts.defineClassAndGetConstructor(ClassLoaderForClassArtifacts.java:74)
        at org.codehaus.groovy.runtime.callsite.CallSiteGenerator.compilePojoMethod(CallSiteGenerator.java:250)
        at org.codehaus.groovy.reflection.CachedMethod.createPojoMetaMethodSite(CachedMethod.java:267)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.createCachedMethodSite(PojoMetaMethodSite.java:162)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.createPojoMetaMethodSite(PojoMetaMethodSite.java:151)
        at groovy.lang.MetaClassImpl.createPojoCallSite(MetaClassImpl.java:3408)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createPojoSite(CallSiteArray.java:131)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:165)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
        at org.codehaus.groovy.ast.builder.AstBuilder.buildFromString(AstBuilder.groovy:102)
        at org.codehaus.groovy.ast.builder.AstBuilder.buildFromString(AstBuilder.groovy)
        at HelloWorld.main(HelloWorld.java:20)
```

That is with `--allow-incomplete-classpath`. If I remove that then it won't build with the error:
```
Reason: java.lang.ClassNotFoundException: groovy.lang.GroovyCodeSource$getName
```
I don't quite understand this error as `GroovyCodeSource` is on the classpath, though `getName` is just a regular public method and non an inner class or anything. So I bet is related to how groovy changes the internals of it all.

**Steps to reproduce the issue**
Please include both build steps as well as run steps
1. **[e.g.: git clone --depth 1 https://git.myrepo.com/projectone ]**
2. **[e.g.: mvn clean package]**

**Describe GraalVM and your environment:**
 - GraalVM version: CE 21.3
 - JDK major version: 11
 - OS: Arch Linux
 - Architecture: AMD64

**More details**
The two config files included are from agent runs of the actual jar I am wrapping. So the `reflect-config.json` has a log of the Groovy classes in it including the `GroovyCodeSource$getName` reference above. The `resource-config.json` has a `"pattern":"\\QMETA-INF/dgminfo\\E"`. If I remove this then I get a `org.codehaus.groovy.runtime.metaclass.MissingMethodExceptionNoStack: No signature of method: java.lang.String.asBoolean() is applicable for argument types: () values: []` which  again looks obviously Groovy related as there is no `asBoolean` in `String` of course, that is a Groovy thing.

Maven output:
```
$ mvn clean package
[INFO] Scanning for projects...
[WARNING]
[WARNING] Some problems were encountered while building the effective model for
com.example:hello-world:jar:1.0.0-SNAPSHOT
[WARNING] 'build.plugins.plugin.version' for org.apache.maven.plugins:maven-compiler-plugin is missing. @ line 51,
column 15
[WARNING]
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING]
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING]
[INFO]
[INFO] ----------------------< com.example:hello-world >-----------------------
[INFO] Building hello-world 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ hello-world ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ hello-world ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] Copying 2 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ hello-world ---
[INFO] Changes detected - recompiling the module!
[WARNING] File encoding has not been set, using platform encoding UTF-8, i.e. build is platform dependent!
[INFO] Compiling 1 source file to
/home/collin/dev/work/hello-world-groovy/target/classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ hello-world ---
[WARNING] Using platform encoding (UTF-8 actually) to copy filtered resources, i.e. build is platform dependent!
[INFO] skip non existing resourceDirectory
/home/collin/dev/work/hello-world-groovy/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ hello-world ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ hello-world ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ hello-world ---
[INFO] Building jar:
/home/collin/dev/work/hello-world-groovy/target/hello-world-1.0.0-SNAPSHOT.jar
[INFO]
[INFO] --- native-maven-plugin:0.9.7.1:build (default) @ hello-world ---
[INFO] ImageClasspath Entry: org.codehaus.groovy:groovy:jar:2.4.21:compile
(file:///home/collin/.m2/repository/org/codehaus/groovy/groovy/2.4.21/groovy-2.4.21.jar)
[INFO] ImageClasspath Entry: com.example:hello-world:jar:1.0.0-SNAPSHOT
(file:///home/collin/dev/work/hello-world-groovy/target/hello-world-1.0.0-SNAPSHOT.jar)
[INFO] Executing: /home/collin/.sdkman/candidates/java/21.3.0.r11-grl/bin/native-image -cp
/home/collin/.m2/repository/org/codehaus/groovy/groovy/2.4.21/groovy-2.4.21.jar:/home/collin/dev/work/hello-world-groovy/target/hello-world-1.0.0-SNAPSHOT.jar
--no-fallback --allow-incomplete-classpath --native-image-info --verbose
-H:ConfigurationFileDirectories=/home/collin/dev/work/hello-world-groovy/target/classes/native-image-configs/
-H:Class=HelloWorld -H:Name=hello-world
Executing [
/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/bin/java \
-XX:+UseParallelGC \
-XX:+UnlockExperimentalVMOptions \
-XX:+EnableJVMCI \
-Dtruffle.TrustAllTruffleRuntimeProviders=true \
-Dtruffle.TruffleRuntime=com.oracle.truffle.api.impl.DefaultTruffleRuntime \
-Dgraalvm.ForcePolyglotInvalid=true \
-Dgraalvm.locatorDisabled=true \
-Dsubstratevm.IgnoreGraalVersionCheck=true \
--add-exports=java.base/com.sun.crypto.provider=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.event=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.loader=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.logger=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.misc=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.module=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.org.objectweb.asm=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.org.xml.sax.helpers=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.perf=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.ref=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.reflect=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.util.xml.impl=ALL-UNNAMED \
--add-exports=java.base/jdk.internal.util.xml=ALL-UNNAMED \
--add-exports=java.base/sun.invoke.util=ALL-UNNAMED \
--add-exports=java.base/sun.nio.ch=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.annotation=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.reflectiveObjects=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.repository=ALL-UNNAMED \
--add-exports=java.base/sun.reflect.generics.tree=ALL-UNNAMED \
--add-exports=java.base/sun.security.jca=ALL-UNNAMED \
--add-exports=java.base/sun.security.provider=ALL-UNNAMED \
--add-exports=java.base/sun.security.util=ALL-UNNAMED \
--add-exports=java.base/sun.text.spi=ALL-UNNAMED \
--add-exports=java.base/sun.util.calendar=ALL-UNNAMED \
--add-exports=java.base/sun.util.locale.provider=ALL-UNNAMED \
--add-exports=java.base/sun.util.resources=ALL-UNNAMED \
--add-exports=java.xml.crypto/org.jcp.xml.dsig.internal.dom=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.site=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code.stack=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.code=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.common=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.aarch64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.amd64=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot.sparc=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.hotspot=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.meta=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.runtime=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.services=ALL-UNNAMED \
--add-exports=jdk.internal.vm.ci/jdk.vm.ci.sparc=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.events=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.consumer=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.handlers=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal.jfc=ALL-UNNAMED \
--add-exports=jdk.jfr/jdk.jfr.internal=ALL-UNNAMED \
-XX:+UseJVMCINativeLibrary \
-Xss10m \
-Xms1g \
-Xmx14g \
-Duser.country=US \
-Duser.language=en \
-Djava.awt.headless=true \
-Dorg.graalvm.version=21.3.0 \
-Dorg.graalvm.config=CE \
-Dcom.oracle.graalvm.isaot=true \
-Djava.system.class.loader=com.oracle.svm.hosted.NativeImageSystemClassLoader \
-Xshare:off \
-Djdk.internal.lambda.disableEagerInitialization=true \
-Djdk.internal.lambda.eagerlyInitialize=false \
-Djava.lang.invoke.InnerClassLambdaMetafactory.initializeLambdas=false \
-javaagent:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/svm.jar \
-cp \
/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/pointsto.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/objectfile.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/llvm-wrapper-shadowed.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/svm.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/svm-llvm.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/javacpp-shadowed.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/builder/llvm-platform-specific-shadowed.jar
\
--module-path \
/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/truffle/truffle-api.jar \
'com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus' \
-watchpid \
60830 \
-imagecp \
/home/collin/.m2/repository/org/codehaus/groovy/groovy/2.4.21/groovy-2.4.21.jar:/home/collin/dev/work/hello-world-groovy/target/hello-world-1.0.0-SNAPSHOT.jar:/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/library-support.jar
\
-H:Path=/home/collin/dev/work/hello-world-groovy/target \
-H:FallbackThreshold=0 \
-H:+AllowIncompleteClasspath \
-H:+DumpTargetInfo \
-H:ConfigurationFileDirectories=/home/collin/dev/work/hello-world-groovy/target/classes/native-image-configs/
\
-H:Class=HelloWorld \
-H:Name=hello-world \
-H:CLibraryPath=/home/collin/.sdkman/candidates/java/21.3.0.r11-grl/lib/svm/clibraries/linux-amd64
]
[hello-world:60851]    classlist:     682.00 ms,  0.96 GB
[hello-world:60851]        (cap):     381.36 ms,  0.96 GB
Warning: Could not resolve groovy.lang.GroovyCodeSource$getName for reflection configuration. Reason:
java.lang.ClassNotFoundException: groovy.lang.GroovyCodeSource$getName.
Warning: Could not resolve java_lang_String$trim for reflection configuration. Reason: java.lang.ClassNotFoundException:
java_lang_String$trim.
Warning: Could not resolve java_lang_System$currentTimeMillis for reflection configuration. Reason:
java.lang.ClassNotFoundException: java_lang_System$currentTimeMillis.
Warning: Could not resolve java_util_List$add for reflection configuration. Reason: java.lang.ClassNotFoundException:
java_util_List$add.
Warning: Could not resolve javax.activation.ActivationDataFlavor for reflection configuration. Reason:
java.lang.ClassNotFoundException: javax.activation.ActivationDataFlavor.
Warning: Could not resolve org.codehaus.groovy.control.CompilationUnit$addSource for reflection configuration. Reason:
java.lang.ClassNotFoundException: org.codehaus.groovy.control.CompilationUnit$addSource.
Warning: Could not resolve org.codehaus.groovy.control.CompilationUnit$compile$0 for reflection configuration. Reason:
java.lang.ClassNotFoundException: org.codehaus.groovy.control.CompilationUnit$compile$0.
Warning: Could not resolve org.codehaus.groovy.control.CompilePhase$getPhaseNumber for reflection configuration. Reason:
java.lang.ClassNotFoundException: org.codehaus.groovy.control.CompilePhase$getPhaseNumber.
[hello-world:60851]        setup:   1,461.29 ms,  0.96 GB
# Printing compilation-target information to:
# /home/collin/dev/work/hello-world-groovy/target/reports/target_info_20211129_102817.txt
[hello-world:60851]     (clinit):     204.27 ms,  2.92 GB
# Printing native-library information to:
# /home/collin/dev/work/hello-world-groovy/target/reports/native_library_info_20211129_102826.txt
[hello-world:60851]   (typeflow):   2,101.63 ms,  2.92 GB
[hello-world:60851]    (objects):   2,671.97 ms,  2.92 GB
[hello-world:60851]   (features):   3,000.79 ms,  2.92 GB
[hello-world:60851]     analysis:   8,408.93 ms,  2.92 GB
[hello-world:60851]     universe:     805.70 ms,  2.92 GB
[hello-world:60851]      (parse):     560.49 ms,  3.43 GB
[hello-world:60851]     (inline):     612.95 ms,  3.43 GB
[hello-world:60851]    (compile):   5,782.42 ms,  4.41 GB
[hello-world:60851]      compile:   7,935.79 ms,  4.41 GB
[hello-world:60851]        image:   1,237.35 ms,  4.41 GB
[hello-world:60851]        write:     209.67 ms,  4.41 GB
[hello-world:60851]      [total]:  20,881.90 ms,  4.41 GB
# Printing build artifacts to:
# /home/collin/dev/work/hello-world-groovy/target/hello-world.build_artifacts.txt
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  22.206 s
[INFO] Finished at: 2021-11-29T10:28:36-08:00
[INFO] ------------------------------------------------------------------------
```

Try to run binary:
```
$ target/hello-world build.gradle
Hello World
Exception in thread "main" com.oracle.svm.core.jdk.UnsupportedFeatureError: No classes have been predefined during the image build to load from bytecodes at runtime.
        at com.oracle.svm.core.util.VMError.unsupportedFeature(VMError.java:87)
        at com.oracle.svm.core.hub.PredefinedClassesSupport.throwNoBytecodeClasses(PredefinedClassesSupport.java:76)
        at com.oracle.svm.core.hub.PredefinedClassesSupport.loadClass(PredefinedClassesSupport.java:130)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:269)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts.define(ClassLoaderForClassArtifacts.java:43)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts$1.run(ClassLoaderForClassArtifacts.java:76)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts$1.run(ClassLoaderForClassArtifacts.java:74)
        at java.security.AccessController.doPrivileged(AccessController.java:87)
        at org.codehaus.groovy.reflection.ClassLoaderForClassArtifacts.defineClassAndGetConstructor(ClassLoaderForClassArtifacts.java:74)
        at org.codehaus.groovy.runtime.callsite.CallSiteGenerator.compilePojoMethod(CallSiteGenerator.java:250)
        at org.codehaus.groovy.reflection.CachedMethod.createPojoMetaMethodSite(CachedMethod.java:267)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.createCachedMethodSite(PojoMetaMethodSite.java:162)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.createPojoMetaMethodSite(PojoMetaMethodSite.java:151)
        at groovy.lang.MetaClassImpl.createPojoCallSite(MetaClassImpl.java:3408)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createPojoSite(CallSiteArray.java:131)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.createCallSite(CallSiteArray.java:165)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:120)
        at org.codehaus.groovy.ast.builder.AstBuilder.buildFromString(AstBuilder.groovy:102)
        at org.codehaus.groovy.ast.builder.AstBuilder.buildFromString(AstBuilder.groovy)
        at HelloWorld.main(HelloWorld.java:20)
```

**Comment by rodrigar-mx** [comment]
Hi @collinpeters. Thanks for sharing the issue. We will take a look into it and get back to you.

**Comment by rodrigar-mx** [comment]
Hi @collinpeters. I tried to run it and did not get any issue if I generate the config files myself. These were the steps I followed:
1. In a new empty directory I copied HelloWorld.java and build.gradle files from https://github.com/collinpeters/graalvm-groovy
2. I downloaded groovy.jar lib from https://mvnrepository.com/artifact/org.codehaus.groovy/groovy/3.0.9
3. Compile: _javac -cp groovy-3.0.9.jar:. HelloWorld.java_
4. Run tracing agent: _java -agentlib:native-image-agent=config-output-dir=META-INF/native-image -cp groovy-3.0.9.jar:. HelloWorld_
5. Build native image: _native-image --no-fallback --allow-incomplete-classpath --report-unsupported-elements-at-runtime -cp groovy-3.0.9.jar:. HelloWorld_
6. _./helloworld_

Graalvm version used:
```
openjdk version "11.0.13" 2021-10-19
OpenJDK Runtime Environment GraalVM CE 21.3.0 (build 11.0.13+7-jvmci-21.3-b05)
OpenJDK 64-Bit Server VM GraalVM CE 21.3.0 (build 11.0.13+7-jvmci-21.3-b05, mixed mode, sharing)
```
